Fix various bugs in subclass input in export

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163770

This adds basic support for subclass inputs in export (specifically for non-strict). I had to make fakify little more complicated which risks further divergence from dynamo fakification. But dynamo one is so complex, so i feel it is better to do this way. Also improved fake mode detection logic to recursively look into subclass inner tensors. 

Differential Revision: [D83156489](https://our.internmc.facebook.com/intern/diff/D83156489)

tugsbayasgalan committed:
Fix various bugs in subclass input in export

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D83156489).

angelayi commented on torch/_export/non_strict_utils.py:
```diff
@@ -199,26 +199,71 @@ def fakify(
+
+    if is_traceable_wrapper_subclass(t):
+        # Get symbolic context for outer tensor
+        outer_context = _create_symbolic_context_for_tensor(
```
what does specifying dynamic shapes for tensor subclasses look like? could you add a test or add some casing here to error if we try to have dynamic shapes with tensor subclasses?

avikchaudhuri reviewed (commented):
lots of questions, mostly for my own education, although more comments would be nice

tugsbayasgalan committed:
Update on "Fix various bugs in subclass input in export"


Differential Revision: [D83156489](https://our.internmc.facebook.com/intern/diff/D83156489)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D83156489).

tugsbayasgalan commented on torch/_guards.py:
```diff
@@ -1151,6 +1156,18 @@ def detect_fake_mode(inputs: Any = None) -> Optional[FakeTensorMode]:
     for i, flat_input in enumerate(flat_inputs):
         if isinstance(flat_input, FakeTensor):
             fake_modes.append((flat_input.fake_mode, "fake tensor input", i))
+        if is_traceable_wrapper_subclass(flat_input):
```
Yeah FakeTensor doens't have flatten/unflatten.

tugsbayasgalan committed:
Update on "Fix various bugs in subclass input in export"


Differential Revision: [D83156489](https://our.internmc.facebook.com/intern/diff/D83156489)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D83156489).

tugsbayasgalan committed:
Update on "Fix various bugs in subclass input in export"


This adds basic support for subclass inputs in export (specifically for non-strict). I had to make fakify little more complicated which risks further divergence from dynamo fakification. But dynamo one is so complex, so i feel it is better to do this way. Also improved fake mode detection logic to recursively look into subclass inner tensors. 

Differential Revision: [D83156489](https://our.internmc.facebook.com/intern/diff/D83156489)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix various bugs in subclass input in export"


This adds basic support for subclass inputs in export (specifically for non-strict). I had to make fakify little more complicated which risks further divergence from dynamo fakification. But dynamo one is so complex, so i feel it is better to do this way. Also improved fake mode detection logic to recursively look into subclass inner tensors. 

Differential Revision: [D83156489](https://our.internmc.facebook.com/intern/diff/D83156489)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D83156489).

Camyll commented:
@pytorchmergebot merge
